## Evolutionary Multi-objective Optimisation (EMO) algorithms:

|NSGA-II|SPEA2|GDE3|IBEA|
| ---| --- | --- | --- |
|Non-Dominated Sorting Genetic Algorithm (2nd version)|Strength Pareto Evolutionary Algorithm 2|The third version of Generalized Differential Evolution algorithm| Indicator-based evolutionary algorithm|
|A fast non-dominated sorting approach to achieve solutions as close as possible to pareto-optimal solutions|It is based on the concept of Pareto domination for fitness allocation and selection operations|Randomly select the initial parameter values uniformly on the intervals| Incorporate an arbitrary performance indicator in its fitness function, using binary permirmance indicator to compare the diffence between two solutions|
|A fast crowded distance estimation procedure|Density based on the k-th nearest neighbor|Stochastic, population-based optimisation algorithm|Based on a binary predifined indicator to compare the difference between two solutions|
|A simple crowded comparison operator|Its genetic evolution operation has strong randomness|Designed for any number of objectives and constraints without introducing any extra control parameters|IBEA with the hypervolume measures have theoritical support and search ability|
|The best solutions in the current population are preserved and used in the next generation|Improve SPEA making sure extreme points are preserved| Has two control parameters: mutation amplification factor and crossover rate|Uses both recombination and mutation operators to generate offspring|
|It is a fast algorithm which does not demand high computational cost|Computationally expensive fitness and density calculation|uses a growing population and nondominated sorting with pruning of non-dominated solutions to decrease the population size at the end of each generation|Hypervolume calculation needs long computation time|




## Stochastic Optimisation

.
<img src="https://drbo.bg/wp-content/uploads/2018/05/cube-1979772_1280-768x403.jpg" width="300" height="240" align="center"/>

* Collection of methods for minimising (maximising) an objective function when randomness is present, ussually enters the problem in two ways:
    - Cost function 
    - Constrains set 

* Structural assumptions, such as limits on the size of the decision and outcome spaces, or convexity, are needed to make problems tractable.


## Testable Hypothesis:

* According to Carvalho and Sichman (2018), IBEA will outperform NSGA-II, SPEA2 and GDE3
* Based on the characteristics of this problem, we could say that the ZDT test suite is the most suitable one for the Welded beam Design problem as this has two objectives problem. 
* Two major problems must be addressed when an evolutionary algorithm is applied to multi-objective optimisation: 
    - How to accomplish fitness assignment and selection, in order to guide the search towards the Pareto -optimal set.
    - How to maintain a diverse population in order to prevent premature convergence and archive a well distribute trade-off front. 


## Synthetic Experiment Design & Execution

### Performance metric for comparison: Hypervolume Indicator

- Popular method to compare Pareto sets by measuring the volume of the dominated portion of objective space. 
- Suitable performance metric for our experiment because:
  - Balanced measurement by assessing proximity, diversity and pertinence. 
  - Assessing the results is a straightforward procedure, higher hypervolume values are more desirable. 
- The recommended reference point for our chosen synthetic test problems is (11,11). 


### Statistical Comparison Design

- **Sample Size sufficiency test**: investigate the relationship between sample size and the Stand Error of the Mean (SEM) to determine an appropriate sample size for the test suite problems.
  - **Our chosen SEM level**: 0.005
- Mean Hypervolume scores to determine the best performing algorithm
- **Significance Test**: Wilcoxon signed-rank test to determine if the results are significant or likely from the same distribution


In [ ]:
import numpy as np                   # for multi-dimensional containers
import pandas as pd                  # for DataFrames
import plotly.graph_objects as go    # for data visualisation
import plotly.io as pio              # to set shahin plot layout
import platypus as plat              # multi-objective optimisation framework
from scipy import stats

In [ ]:
#codes used to determine sufficient sample size
problem = [plat.ZDT1]
algorithms = [plat.NSGAII]

results_1 = plat.experiment(algorithms, problem, nfe=10000, seeds=150)

hyp = plat.Hypervolume(minimum=[0, 0], maximum=[11, 11])

hyp_result_1 = plat.calculate(results_1, hyp)

hyp_result_1 = hyp_result_1['NSGAII']['ZDT1']['Hypervolume']

SEM = []

for sample_size in range(3,len(hyp_result_1)):
    SEM.append(stats.sem(hyp_result_1[0:sample_size]))
    
algorithms_2 = [plat.SPEA2]

results_2 = plat.experiment(algorithms_2, problem, nfe=10000, seeds=150)

hyp_result_2 = plat.calculate(results_2, hyp)

hyp_result_2 = hyp_result_2['SPEA2']['ZDT1']['Hypervolume']

SEM_2 = []


for sample_size in range(3,len(hyp_result_2)):
    SEM_2.append(stats.sem(hyp_result_2[0:sample_size]))
    
algorithms_3 = [plat.GDE3]

results_3 = plat.experiment(algorithms_3, problem, nfe=10000, seeds=150)

hyp_result_3 = plat.calculate(results_3, hyp)

hyp_result_3 = hyp_result_3['GDE3']['ZDT1']['Hypervolume']

SEM_3 = []


for sample_size in range(3,len(hyp_result_3)):
    SEM_3.append(stats.sem(hyp_result_3[0:sample_size]))
    
algorithms_4 = [plat.IBEA]

results_4 = plat.experiment(algorithms_4, problem, nfe=10000, seeds=150)

hyp_result_4 = plat.calculate(results_4, hyp)

hyp_result_4 = hyp_result_4['IBEA']['ZDT1']['Hypervolume']

SEM_4 = []


for sample_size in range(3,len(hyp_result_4)):
    SEM_4.append(stats.sem(hyp_result_4[0:sample_size]))
    
level=0.005 #level to decide the sample size
xlevel = np.linspace(0,100) #for plotting the level line
ylevel = 0*xlevel+level

fig_1 = go.Figure(layout=dict(xaxis=dict(title='Sample Size - NSGA - SPEA2 - GDE3 - IBEA ZDT1'), yaxis=dict(title='Standard Error of the Mean')))
fig_1.add_scatter(x=list(range(3,len(hyp_result_1))), y=SEM,name='NSGAII')
fig_1.add_scatter(x=list(range(3,len(hyp_result_2))), y=SEM_2,name='SPEA2')
fig_1.add_scatter(x=list(range(3,len(hyp_result_3))), y=SEM_3,name='GDE3')
fig_1.add_scatter(x=list(range(3,len(hyp_result_4))), y=SEM_4,name='IBEA')
fig_1.add_scatter(x=xlevel, y=ylevel, name='level line')
fig_1.show()

### Algorithm Configurations

- **Sample Size**: 50 for all, 120 for ZDT4 due to IBEA performance 
- **Number of function evaluations (nfe)**: 10,000 for all, 25000 for ZDT4 as per recommended literature (Chase N, et al. 2009) )
- **Genetic operators**: Default parameters to be used for each algorithm

In [ ]:
from IPython.display import display, HTML

def multi_table(table_list):
    '''
    '''
    return HTML(
    '<table><tr style="background-color:white;">' +
    ''.join(['<td>' + table._repr_html_() + '</td>' for table in table_list]) +
    '</tr></table>'
    )

In [ ]:
#Results Generation example code:

problem = [plat.ZDT1,plat.ZDT2,plat.ZDT3,plat.ZDT6]
algorithms = [plat.NSGAII, plat.SPEA2, plat.GDE3, plat.IBEA]

In [ ]:
results = plat.experiment(algorithms, problem, nfe=10000, seeds=50)

In [ ]:
hyp = plat.Hypervolume(minimum=[0, 0], maximum=[11, 11])

In [ ]:
hyp_result = plat.calculate(results, hyp)

In [ ]:
df_hyp_results = pd.DataFrame(index = hyp_result.keys())

for key_algorithm, algorithm in hyp_result.items():
    for key_problem, problem in algorithm.items():
        for hypervolumes in problem['Hypervolume']:
            df_hyp_results.loc[key_algorithm,key_problem] = np.mean(hypervolumes)
            

In [ ]:
#Seperate problem parameters

problem_2 = plat.ZDT4

In [ ]:
results_2 = plat.experiment(algorithms, problem_2, nfe=25000, seeds=120)

In [ ]:
hyp_result_2 = plat.calculate(results_2, hyp)

In [ ]:
df_hyp_results_2 = pd.DataFrame(index = hyp_result_2.keys())

for key_algorithm, algorithm in hyp_result_2.items():
    for key_problem, problem in algorithm.items():
        for hypervolumes in problem['Hypervolume']:
            df_hyp_results_2.loc[key_algorithm,key_problem] = np.mean(hypervolumes)
            

### Results Generation

In [ ]:
multi_table([df_hyp_results.transpose(), df_hyp_results_2.transpose()])

#### How do these results compare to our hypothesis? 

- **Hypothesis**: Based on literature Carvalho and Sichman (2018), whom conducted similiar benchmark tests on different synthetic test problems, but with same algorithms as we have chosen, we predict that IBEA will be the strongest performing genetic algorithm. In addition we believe GDE3 will perform strongly, as it is a more modern algorithm with similiar high performance seen in literature. 



### Statistical Comparison: _Statistical Significance Test_

In [ ]:
algorithms = ['NSGAII','SPEA2', 'GDE3','IBEA']
problems = ['ZDT1','ZDT2','ZDT3','ZDT6']

df_hyp_wilcoxon = pd.DataFrame(index = [algorithms[1]])

for key_problem in problems:
    s, p = stats.wilcoxon(hyp_result[algorithms[0]][key_problem]['Hypervolume'],
                          hyp_result[algorithms[1]][key_problem]['Hypervolume'])
    df_hyp_wilcoxon.loc[algorithms[1],key_problem] = p

In [ ]:
problems = ['ZDT1','ZDT2','ZDT3','ZDT6']

df_hyp_wilcoxon_2 = pd.DataFrame(index = [algorithms[2]])


for key_problem in problems:
    s, p = stats.wilcoxon(hyp_result[algorithms[0]][key_problem]['Hypervolume'],
                          hyp_result[algorithms[2]][key_problem]['Hypervolume'])
    df_hyp_wilcoxon_2.loc[algorithms[2],key_problem] = p
            

In [ ]:
problems = ['ZDT1','ZDT2','ZDT3','ZDT6']

df_hyp_wilcoxon_3 = pd.DataFrame(index = [algorithms[3]])


for key_problem in problems:
    s, p = stats.wilcoxon(hyp_result[algorithms[0]][key_problem]['Hypervolume'],
                          hyp_result[algorithms[3]][key_problem]['Hypervolume'])
    df_hyp_wilcoxon_3.loc[algorithms[3],key_problem] = p
            

In [ ]:
problems_2 = ['ZDT4']

df_hyp_wilcoxon_4 = pd.DataFrame(index = [algorithms[1]])


for key_problem in problems_2:
    s, p = stats.wilcoxon(hyp_result_2[algorithms[0]][key_problem]['Hypervolume'],
                          hyp_result_2[algorithms[1]][key_problem]['Hypervolume'])
    df_hyp_wilcoxon_4.loc[algorithms[1],key_problem] = p
            

In [ ]:
problems_2 = ['ZDT4']

df_hyp_wilcoxon_5 = pd.DataFrame(index = [algorithms[2]])


for key_problem in problems_2:
    s, p = stats.wilcoxon(hyp_result_2[algorithms[0]][key_problem]['Hypervolume'],
                          hyp_result_2[algorithms[2]][key_problem]['Hypervolume'])
    df_hyp_wilcoxon_5.loc[algorithms[2],key_problem] = p

In [ ]:
problems_2 = ['ZDT4']

df_hyp_wilcoxon_6 = pd.DataFrame(index = [algorithms[3]])


for key_problem in problems_2:
    s, p = stats.wilcoxon(hyp_result_2[algorithms[0]][key_problem]['Hypervolume'],
                          hyp_result_2[algorithms[3]][key_problem]['Hypervolume'])
    df_hyp_wilcoxon_6.loc[algorithms[3],key_problem] = p

In [ ]:
multi_table([df_hyp_wilcoxon.transpose(),df_hyp_wilcoxon_2.transpose(),df_hyp_wilcoxon_3.transpose()])

In [ ]:
multi_table([df_hyp_wilcoxon_4.transpose(),df_hyp_wilcoxon_5.transpose(),df_hyp_wilcoxon_6.transpose()])

#### Conclusions & further discussion:

- GDE3 performed best in 5 out of 6 test suite problems, therefore the best performing algorithm in the synthentic test suite problems.
- NSGA-II performed best in ZDT3, IBEA performed 2nd best in ZDT1, ZDT3 and ZDT6
- SPEA2 performed the poorest
- The results were statistically significant using NSGA-II as a baseline, apart from ZDT2 with SPEA2 and IBEA
- We decided to use GDE3 for the real-world problem.

### Welded Beam Design Problem


* Problem's description
* Solving the problem with GDE3
* Picking best solutions
* Conclusion






### Welded Beam Design:

**The beam supports a load P at a distance L from the substrate. The beam is welded onto the substrate with upper and lower welds, each of length l and thickness h. The beam has a rectangular cross-section, width b, and height t. The material of the beam is steel.** 


![alt text](https://i.imgur.com/n75MRlP.png)

###### Figure reference: Ray and Liew (2002). A Swarm Metaphor for Multiobjective Design Optimization. Available at: https://i.imgur.com/n75MRlP.png


### The welded beam design problem has:
 - **Four real-parameter variables**:
    - x(1)=h , the thickess of the welds
    - x(2)=l , the length of the welds
    - x(3)=t , the height of the beam
    - x(4)=b , the widht of the beam 
 - **Four non-linear constraints**:
    - Shear stress
    - Bending stress
    - The buckling load
    - Deflection
     
 

In [ ]:
# Definiton of the Welded Beam problem

from platypus import Problem, Real

class WeldedBeam(Problem):

    def __init__(self):
        super(WeldedBeam, self).__init__(4, 2, 4)
        self.types[:] = [Real(0.125, 5.0), Real(0.1, 10.0),Real(0.1, 10.0),Real(0.125, 5.0)]
        self.constraints[:] = "<=0"
    
    def evaluate(self, solution):
        x = solution.variables
        f1 = 1.10471 * x[0] ** 2 * x[1] + 0.04811 * x[2] * x[3] * (14.0 + x[1])
        f2 = 2.1952 / (x[3] * x[2] ** 3)

        P = 6000
        L = 14
        t_max = 13600
        s_max = 30000

        R =np.sqrt(0.25 * (x[1] ** 2 + (x[0] + x[2]) ** 2))
        M = P * (L + x[1] / 2)
        J = 2 *np.sqrt(0.5) * x[0] * x[1] * (x[1] ** 2 / 12 + 0.25 * (x[0] + x[2]) ** 2)
        t1 = P / (np.sqrt(2) * x[0] * x[1])
        t2 = M * R / J
        t =np.sqrt(t1 ** 2 + t2 ** 2 + t1 * t2 * x[1] / R)
        s = 6 * P * L / (x[3] * x[2] ** 2)
        P_c = 64746.022 * (1 - 0.0282346 * x[2]) * x[2] * x[3] ** 3

        g1 = (1 / t_max) * (t - t_max)
        g2 = (1 / s_max) * (s - s_max)
        g3 = (1 / (5 - 0.125)) * (x[0] - x[3])
        g4 = (1 / P) * (P - P_c)
        solution.objectives[:] = [f1, f2]
        solution.constraints[:] = [g1,g2,g3,g4]
        
# Reference: https://github.com/msu-coinlab/pymop/blob/master/pymop/problems/welded_beam.py

In [ ]:
# run the algorithm

nef=10000 #the number of function evaluations
problem=WeldedBeam() 
algorithm = plat.GDE3(WeldedBeam())
algorithm.run(nef)

In [ ]:
#getting the results

objective_values = np.empty((0, 2))
solutions = []

for solution in algorithm.result:
    solution.evaluate()
    solutions.append(solution)
    y = solution.objectives
    objective_values = np.vstack([objective_values, y])
    
# convert to DataFrame
objective_values = pd.DataFrame(objective_values, columns=['f1','f2'])

#print the restults 

objective_values


f1        f2
0   24.712374  0.000672
1   13.611677  0.001255
2   12.594857  0.001361
3   12.405863  0.001390
4   24.784815  0.000666
..        ...       ...
95   3.880289  0.005260
96   3.465052  0.005906
97   4.290115  0.004723
98  36.596274  0.000439
99   3.446015  0.006019

[100 rows x 2 columns]

In [ ]:
#Show the Figure 

fig = go.Figure(layout=dict(xaxis=dict(title='Cost '),yaxis=dict(title='Deflection')))
fig.add_scatter(x=objective_values.f1, y=objective_values.f2, mode='markers',name='solutions')

fig.show()


### Reference point: (40,0.009)

In [ ]:
#Reference point
RefPoint=[40,0.009]

In [ ]:
hyp = plat.indicators.Hypervolume(minimum=[0, 0], maximum=RefPoint)

In [ ]:
#Show the Figure 

fig = go.Figure(layout=dict(xaxis=dict(title='Cost '),yaxis=dict(title='Deflection')))
fig.add_scatter(x=objective_values.f1, y=objective_values.f2, mode='markers',name='solutions')
fig.add_scatter(x=[0,RefPoint[0],RefPoint[0]], y=[RefPoint[1],RefPoint[1],0],name='Reference point',)

fig.show()

### Calculation of the explicit hypervolume contribution of each solution in the population

In [ ]:
HV = hyp.calculate(solutions)
CHV = np.empty((0, 1))

for i in range(100):
    solutions_subset = solutions[:i] + solutions[i+1:]
    CHV = np.vstack([CHV, HV - hyp.calculate(solutions_subset)])
    objective_values = pd.DataFrame(objective_values, columns=['f1','f2'])

CHV = pd.DataFrame(CHV, columns=['CHV'])
CHV=CHV.sort_values(by='CHV',ascending=False)

### The best solutions

- **4 represents the number of picked solutions**:


In [ ]:
#number of picked solutions
noS=4

In [ ]:
# Using loop to print best solutions

for i in range(0,noS):
  solutionIndex= CHV.index[i]
  print(objective_values.loc[[solutionIndex]])
  
  # Add a solution with large marker
  fig.add_trace(
    go.Scatter(
        name=f'solution {solutionIndex}',
        mode='markers',
        x=[objective_values.iloc[solutionIndex].f1],
        y=[objective_values.iloc[solutionIndex].f2],
        marker=dict(
            color='Red',
            size=10,
            line=dict(
                color='Purple',
                width=2
            )
        ),
        
    )
  )
fig.show()


 f1        f2
97  4.290115  0.004723
          f1        f2
99  3.446015  0.006019
          f1        f2
96  3.465052  0.005906
          f1        f2
94  3.901169  0.005165

### References:

* Deb, Kalyanmoy, J. Sundar, Udaya Bhaskara Rao N, and Shamik Chaudhuri. Reference Point Based Multi-Objective Optimization Using Evolutionary Algorithms. International Journal of Computational Intelligence Research, Vol. 2, No. 3, 2006, pp. 273–286. Available at https://www.softcomputing.net/ijcir/vol2-issu3-paper4.pdf

* Ray, T., and K. M. Liew. A Swarm Metaphor for Multiobjective Design Optimization. Engineering Optimization 34, 2002, pp.141–153.

* Deb, Kalyanmoy, J. Sundar, Udaya Bhaskara Rao N, and Shamik Chaudhuri. Reference Point Based Multi-Objective Optimization Using Evolutionary Algorithms. International Journal of Computational Intelligence Research, Vol. 2, No. 3, 2006, pp. 273–286.

* Ali R. Yildiz. Comparison of evolutionary-based optimization algorithms for structuraldesign optimization. Engineering Applications of Artificial Intelligence. 2012.

* Hamidreza, Eskandari,G. Christopher, and Lamont Gary. FastPGA: A Dynamic Population Sizing Approach for Solving Expensive Multiobjective Optimization Problems. 2007.

* Zhao Liang. Performance of multi-objectve algorithms on varying test problem features using the WFG toolkit. 2007

* Y. Yusoff, M. S. Ngadiman, and A. M. Zain, “Overview of NSGA-II for optimizing machining process parameters,” in Procedia Engineering, 2011. K. Deb, A. Pratap, S. Agarwal, and T. Meyarivan, “A fast and elitist multiobjective genetic algorithm: NSGA-II,” IEEE Trans. Evol. Comput., 2002.

* K. Deb, A. Pratap, S. Agarwal, and T. Meyarivan, “A fast and elitist multiobjective genetic algorithm: NSGA-II,” IEEE Trans. Evol. Comput., 2002.

* Kukkonen, S. and Lampinen, J., 2005, September. GDE3: The third evolution step of generalized differential evolution. In 2005 IEEE congress on evolutionary computation (Vol. 1, pp. 443-450). IEEE.

* E. Zitzler, K. Deb, and L. Thiele. Comparison of Multiobjective Evolutionary Algorithms: Empirical Results. Evolutionary Computation, 8(2):173-195, 2000 

* Rostami, S., O’Reilly, D., Shenfield, A. and Bowring, N., 2015. A novel preference articulation operator for the evolutionary multi-objective optimisation of classifiers in concealed weapons detection. Information Sciences, 295, pp.494-520.

* Chase, N., Rademacher, M., Goodman, E., Averill, R. and Sidhu, R., 2009. A benchmark study of multi-objective optimization methods. BMK-3021, Rev, 6, pp.1-24.